### PR TITLE
Ad-hoc fix for enqueuing attach attempt when DISCONNECTED.

### LINKED ISSUE #189: Channel attach fails in the DISCONNECTED state
Channel attach operations when in a `DISCONNECTED` state fail with an error: `attempted to attach channel to inactive connection`.

### PR DESCRIPTION
**Pull request by tcard**
To be converted to a proper spec-oriented test when ported to the 1.2 branch.

Fixes #189.

### CODE REVIEW

File: ably/realtime_channel.go [hunk 1]
```diff
@@ -167,9 +167,37 @@ func (c *RealtimeChannel) mayAttach(result, checkActive bool) (Result, error) {
...
+
+	// RTL4i
+	case StateConnInitialized,
+		StateConnConnecting,
+		StateConnDisconnected:
+
+		return goWaiter(func() error {
+			connected := make(chan State, 1)
+			c.client.Connection.On(connected, StateConnConnected)
+			<-connected
```
Comment (gernest) [inline ably/realtime_channel.go:186]: This will block forever when `StateConnConnected` isn't fired.

File: ably/realtime_channel.go [hunk 2]
```diff
@@ -167,9 +167,37 @@ func (c *RealtimeChannel) mayAttach(result, checkActive bool) (Result, error) {
...
-		return nil, c.state.set(StateChanFailed, errAttach)
+
+	switch c.client.Connection.State() {
+	// RTL4b
+	case StateConnClosed,
+		StateConnClosing,
+		StateConnFailed:
+		return nil, newError(80000, errAttach)
+
+	// RTL4i
```
Comment (gernest) [inline ably/realtime_channel.go:178]: can't find this on the spec.

**Review by gernest — CHANGES_REQUESTED** [review]
Left a comment.

Also can you point me on the spec on expected behaviour for this scenario?

File: ably/realtime_channel.go (hunk 1 above)
Comment (tcard) [reply ably/realtime_channel.go]: Good catch! Fixed at 71e958c.

File: ably/realtime_channel.go (hunk 2 above)
Comment (tcard) [reply ably/realtime_channel.go:178]: [RTL4i](https://docs.ably.io/client-lib-development-guide/features/#RTL4i)

File: ably/realtime_channel.go [hunk 3]
```diff
@@ -167,9 +167,40 @@ func (c *RealtimeChannel) mayAttach(result, checkActive bool) (Result, error) {
...
+	case StateConnInitialized,
```
Comment (gernest) [inline ably/realtime_channel.go:179]: >If the connection state is CONNECTING or DISCONNECTED, do the operation once the connection state is CONNECTED

I don't see spec mentioning  `StateConnInitialized` why is it here?

**Review by gernest — CHANGES_REQUESTED** [review]
Left a comment. Is it possible to port the changes to `integration/1.2` branch as well?

File: ably/realtime_channel.go (hunk 3 above)
Comment (tcard) [reply ably/realtime_channel.go]: True, [RTL4b](https://docs.ably.io/client-lib-development-guide/features/#RTL4b) says that would be an error. Fixed at 6fb73b8. (Tests to ensure compliance will follow in the 1.2 branch.)

**Comment by tcard** [comment]
> Is it possible to port the changes to `integration/1.2` branch as well?

Yes, will do once merged. We should do this with every change on master.

**Review by gernest — APPROVED** [review]
LGTM

**Pull request merged**